src/Microsoft.Framework.ApplicationHost2/Program.cs

```diff
@@ -22,9 +26,111 @@ public Program(IAssemblyLoaderContainer container, IApplicationEnvironment envir
+            app.VersionOption("--version", GetVersion());
+            var runCmd = app.Command("run", c =>
+                {
+                    // We don't actually execute "run" command here
+                    // We are adding this command for the purpose of displaying correct help information
+                    c.Description = "Run application";
+                    c.OnExecute(() =>
+                    {
+                        runCmdExecuted = true;
+                        return 0;
+                    });
+                },
+            addHelpCommand: false,
+            throwOnUnexpectedArg: false);
+            app.Execute(args);
+
+            options = null;
+            outArgs = null;
+            exitCode = 0;
+
+            if (app.IsShowingInformation)
+            {
+                // If help option or version option was specified, exit immediately with 0 exit code
+                return false;
+            }
+            else if (!(app.RemainingArguments.Any() || runCmdExecuted))
+            {
+                // If no subcommand was specified, show error message
+                // and exit immediately with non-zero exit code
+                Console.WriteLine("Please specify the command to run");
+                exitCode = 2;
+                return false;
+            }
+
+            options = new ApplicationHostOptions();
+            options.WatchFiles = optionWatch.HasValue();
+            options.PackageDirectory = optionPackages.Value();
+
+            options.TargetFramework = _environment.RuntimeFramework;
+            options.Configuration = optionConfiguration.Value() ?? _environment.Configuration ?? "Debug";
+            options.ApplicationBaseDirectory = _environment.ApplicationBasePath;
+            var portValue = optionCompilationServer.Value() ?? Environment.GetEnvironmentVariable(EnvironmentNames.CompilationServerPort);
+
+            int port;
+            if (!string.IsNullOrEmpty(portValue) && int.TryParse(portValue, out port))
+            {
+                options.CompilationServerPort = port;
+            }
+
+            var remainingArgs = new List<string>();
+            if (runCmdExecuted)
+            {
+                // Later logic will execute "run" command
+                // So we put this argment back after it was consumed by parser
+                remainingArgs.Add("run");
+                remainingArgs.AddRange(runCmd.RemainingArguments);
+            }
+            else
+            {
+                remainingArgs.AddRange(app.RemainingArguments);
+            }
+
+            if (remainingArgs.Any())
+            {
+                options.ApplicationName = remainingArgs[0];
+                outArgs = remainingArgs.Skip(1).ToArray();
+            }
+            else
+            {
+                outArgs = remainingArgs.ToArray();
+            }
+
+            return true;
+        }
+
+        private static string GetVersion()
```

Comment:
Well, we call this method _exactly_ once, so I'm not sure it's worth allocating the memory :P